Sources/AsyncHTTPClient/RedirectState.swift

```diff
@@ -0,0 +1,134 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import struct Foundation.URL
+import NIOHTTP1
+
+struct RedirectState {
+    /// number of redirects we are allowed to follow
+    private var count: Int
+
+    /// nil if we should ignore cycles, otherwise a set of all visited URLs
+    private var visited: Set<URL>?
```

Comment:
Do we maybe want to make the ignore cycles vs. versus detect cycles more explicit?

Also I think we should record all visited urls anyways.

Reply:
Oh do we want to use String here instead of URL?